# dataModel.R
#
#   +-----------------------------------------------------------------+
#   |                                                                 |
#   |  Do not edit this file! Edit "myDataModel.R" instead.    |
#   |                                                                 |
#   +-----------------------------------------------------------------+
#
# Purpose:
#
# Version: 1.0
#
# Date:    2018  05  10
# Author:  Boris Steipe (boris.steipe@utoronto.ca)
#
# V 1.0    First code 2018
#
# TODO:
#
#
# == HOW TO WORK WITH THIS FILE ================================================
#
#  This file contains scenarios and tasks, we will discuss them in detail in
#  class. Edit profusely, write code, experiment with options, or just play.
#  Especially play.
#
#  If there is anything you don't understand, use R's help system,
#  Google for an answer, or ask. Especially ask. Don't continue if you don't
#  understand what's going on. That's not how it works ...
#
# ==============================================================================


#TOC> ==========================================================================
#TOC>
#TOC>   Section  Title                     Line
#TOC> -----------------------------------------
#TOC>   1        SCENARIO                    46
#TOC>   2        READ DATA                   55
#TOC>   3        ANALYZE THE DATA            98
#TOC>   4        PLOT DATA                  114
#TOC>
#TOC> ==========================================================================


# =    1  SCENARIO  ============================================================

# Our data is becoming more and more high-dimensional, as every biomolecule
# has been observed in many variant states, and has been richly annotated.
# In this unit we will retrieve some annotations for a protein encoded in the
# genomic region we worked with in the sequence analysis unit, we will integrate
# them into a data model, and we will design a visualization of the annotated
# data.

# We wish to create a plot that looks like this:

source("./sampleSolutions/dataModelSampleSolutions-ShowPlot.R")


# =    2  READ DATA  ===========================================================

# Task 2.1: Open coordinates 58,815,001 to 58,915,000 of the hg38 assembly
#           of chromosome 20 in the Ensembl genome browser. What gene is
#           annotated to this region?

# Task 2.2: GNAS is a complex locus with multiple transcripts. Download the
#           transcript coordinates for protein coding genes. Hint: download
#           the data from the corresponding Ensembl gene page.
#           - Save the results page as "ENSG00000087460data.csv"
#           - Read the file into an R data frame called GNAStranscripts

# Task 2.3  Remove all rows from GNAStranscripts that are not protein coding:
#           - what column are we looking at?
#           - what values exist in this column?
#           - how do we subset the data frame to the values we want?
#           - how many transcripts do we have? What are their IDs?
#           - restrict the rows to contain only Ensembl transcripts. How many
#             transcripts are left?

# Task 2.4  Calculate the transcript lengths for all transcripts. Store
#           them in a named vector called "tLengths".
#
# Task 2.5  Find the GNAS page on the intogen cancer driver gene website.
#           Explore the page. To download the mutation distribution, you need
#           to register (databases need records of who uses them to compete
#           for funding.) You can register and download, or use the file
#           "./data/GNAS-distribution-data.tsv" instead.
#           - Read the file into a data frame called "GNASmutations".


# =    3  EXPLORE DATA  ========================================================
#
# Task 3.1  View GNASmutations. What do you see?
#           - How many observations of each transcript?
#           - Are there transcripts that are not in our Ensembl table?
#               (hint: use the %in% operator)
#           - How many of each mutation type? Plot that!
#
#           - Are the reference nucleotides correct for our GRCh38 data?
#






# =    3  A DATA MODEL  ====================================================

# The resulting data is all over the place. We have a table with transcript
# annotations, a derived vector of lengths


# =    3  ANALYZE THE DATA  ====================================================
#
# Task 3.1: Calculate the GC contents
#           - How?

# Task 3.2: Is this value expected? What is the human average?
#           - Find the value. Is it constant throughout the genome?

# Task 3.3: What are the dinucleotide frequencies?
#           - Make a vector of dinucleotides (hint: use paste() )
#           - Get the counts (hint: use table() )

# Task 3.4: What are the expected dinucleotide frequencies?
#           - Can we get them from first principles?
#           - Can we simulate them (hint: use sample())

# =    4  PLOT DATA  ===========================================================
#

# Task 4.1: Plot observed and expected frequencies as a barplot().

# =    4  MORE PRACTICE  =======================================================
#
# Is the observed ratio of missense/nonsense/synonymous variants for GNAS
# similar to what one would expect?

#  -  Write a function that executes a loop N times (for N <- 100000) to create
#     a point mutation randomly in the GNAS gene. Keep track of the
#     number of missense, silent ("synonymous"), and nonsense ("truncating")"
#     mutations you find. Count changes of the start codon and the stop
#     codon as "nonsense".

# Here is a header that specifies the function, its parameters and its value:

evalMut <- function(FA, N) {
    # Purpose: evaluate the distribution of silent, missense and nonsense
    # codon changes in cDNA read from FA for N random mutation trials.
    # Parameters:
    #     FA   chr      Filename of a FASTA formatted sequence file of cDNA
    #                     beginning with a start codon.
    #     N    integer  The number of point mutation trials to perform
    # Value:   list     List with the following elements:
    #                      FA    chr  the input file name
    #                      N     num  number of trials performed
    #                      nSilent    num  the number of silent mutations
    #                      nMissense  num  the number of missense mutations
    #                      nNonsense  num  the number of nonsense mutations

}

#  -  Contrast your findings with the relative frequency of the mutations in
#     each category reported on the IntOGen Web page for GNAS.

#  -  Do you think there is an important difference between the expected
#     categories of mutations (i.e. the stochastic background that you
#     simulated), and categories of mutations that were observed in cancer
#     genomes? How could you quantify that?



# [END]
