# dataModelSampleSolutions.R
#
#   +-----------------------------------------------------------------+
#   |                                                                 |
#   |  Do not edit this file!                                         |
#   |                                                                 |
#   +-----------------------------------------------------------------+
#
# Purpose:  Sample solutions for the Data Model workshop unit.
#
# Version: 1.0
#
# Date:    2018  05  10
# Author:  Boris Steipe (boris.steipe@utoronto.ca)
#
# V 1.0    First code 2018
#
# TODO:
#
# ==============================================================================


#TOC> ==========================================================================
#TOC>
#TOC>   Section  Title                     Line
#TOC> -----------------------------------------
#TOC>   1        READ DATA                   35
#TOC>   2        ANALYZE THE DATA           122
#TOC>   3        PLOT DATA                  213
#TOC>
#TOC> ==========================================================================


# =    1  READ DATA  ===========================================================

# Task 2.1: Open coordinates 58,815,001 to 58,915,000 of the hg38 assembly
#           of chromosome 20 in the Ensembl genome browser. What gene is
#           annotated to this region?
#           - Google finds    https://useast.ensembl.org/index.html
#           - Simply enter "human 20:58815001-58915000" into the search field.

# Task 2.2: GNAS is a complex locus with multiple transcripts. Download the
#           transcript coordinates for protein coding genes. Hint: download
#           the data from the corresponding Ensembl gene page.
#           - click on any of the exons to open a pop-up - all exons lead to the
#             same gene: ENSG00000087460
#             https://useast.ensembl.org/Homo_sapiens/Gene/Summary?db=core;g=ENSG00000087460
#           - Click on "Export Data"
#           - Select "CSV (Comma Separated Values)"
#           - Check ONLY the "Gene information" checkbox
#           - Click on "Next >"
#           - Save the result page as "ENSG00000087460data.csv"
#           - Read the file into an R data frame called GNAStranscripts.

GNAStranscripts <- read.csv("ENSG00000087460data.csv",
                            header = TRUE,
                            stringsAsFactors = FALSE)

head(GNAStranscripts)

# Task 2.3  Remove all rows from GNAStranscripts that are not protein coding.
#           - what column are we looking at?
#           "gene_type"
#
#           - what values exist in this column?
unique(GNAStranscripts$gene_type)

#           - how do we subset the data frame to the values we want?
sel <- which(GNAStranscripts$gene_type == "protein_coding")
GNAStranscripts <- GNAStranscripts[sel, ]

#           - how many transcripts do we have? What are their IDs?
(x <- unique(GNAStranscripts$transcript_id))

#           - restrict the rows to contain only Ensembl transcripts. How many
#             transcripts are left?
sel <- grep("^ENST", GNAStranscripts$transcript_id)
GNAStranscripts <- GNAStranscripts[sel, ]
(x <- unique(GNAStranscripts$transcript_id))

# Task 2.4  Calculate the transcript lengths for all transcripts. Store
#           them in a named vector called "tLengths".
#           - many ways to do this, here's one

# initialize a numeric vector, named with unique transcript ids
x <- unique(GNAStranscripts$transcript_id)
tLengths <- numeric(length(x))
names(tLengths) <- x

# iterate over all transcript ids, sum lengths
for (id in names(tLengths)) {
    sel <- which(GNAStranscripts$transcript_id == id)
    l <- 0
    for (idx in sel) {
        l <- l + (GNAStranscripts$end[idx] - GNAStranscripts$start[idx]) + 1
    }
    tLengths[id] <- l
}

plot(sort(tLengths, decreasing = TRUE))

# Task 2.5  Find the GNAS page on the IntOGen cancer driver gene website.
#             https://www.intogen.org/search?gene=GNAS
#           Explore the page. To download the mutation distribution, you need
#           to register (databases need records of who uses them to compete
#           for funding.) You can register and download, or use the file
#           "./data/GNAS-distribution-data.tsv" instead.
#
#           - Read the file into a data frame called "GNASmutations"
GNASmutations <- read.delim("./data/GNAS-distribution-data.tsv",
                            stringsAsFactors = FALSE)

# =    3  EXPLORE DATA  ========================================================
#
# Task 3.1  View GNASmutations. What do you see?
#           - How many observations of each transcript?
(x <- sort(table(GNASmutations$TRANSCRIPT), decreasing = TRUE))
length(x)

#           - Are there transcripts that are not in our Ensembl table?
#               (hint: use the %in% operator)
#             %in% is one of the most important operators in data analysis,
#             I need it practically every day!
#             Here is a prototype example:
c("C", "A", "B") %in% c("E", "A", "D", "C", "G")  # TRUE TRUE FALSE
c("E", "A", "D", "C", "G") %in% c("C", "A", "B")  # FALSE TRUE FALSE TRUE FALSE
#             - The output is a boolean vector;
#             - it has the length of the first set;
#             - it is TRUE for every element of the first set that
#                 appears in the second set.
#             To count TRUE values, we can use the sum(<logical>) approach.
#             or we can use all() or any().
sum(GNASmutations$TRANSCRIPT %in% GNAStranscripts$transcript_id)
all(GNASmutations$TRANSCRIPT %in% GNAStranscripts$transcript_id)

#           - How many of each mutation type? Plot that!
(x <- sort(table(GNASmutations$MOST_SEVERE), decreasing = TRUE))
length(x)

#           - Are the reference nucleotides correct for our GRCh38 data?
#             - extract a data frame of coordinates and reference alleles
myIntRef <- GNASmutations[ , c("START", "REF")]

#             - remove all rows that are not single nucleotides
myIntRef <- myRef[ (myIntRef$REF %in% c("A", "C", "G", "T")), ]  # or ...
myIntRef <- myIntRef[grep("[ACGT]", myIntRef$REF), ]

#             - look at it. Now remove all duplicates:
sel <- duplicated(myIntRef$START)
myIntRef <- myIntRef[! sel, ]

#             - calculate offset to mySeq position 1
#                 We downloaded from 20:58815001-58915000 - therefore a
#                 position i in myseq is i + 58815000 on chr20.
#
#             - look at the numbers. What's going on?
#
#             Apparently IntOGen uses GRCh37 coordinates!
#             Always confirm and test your data!!!
#
#             - How do we recover?
#             Easy: since everything is in a script, we simply re-run our script
#             on GRCh37 archive data!
#
#             - Let's first rename mySeq
mySeq38 <- mySeq
#
#             - next we find a gene source for GRCh37. We started from ...
#               https://useast.ensembl.org/Homo_sapiens/Info/Index ...
#               ... find "GRCh37 full archive ..." on that page and click "Go".
#               Of course, searching for the same coordinates won't help, so
#               search for GNAS ... and click on "GNAS human gene". Coordinates
#               57,410,001-57,510,000 seem to cover the region. Copy the URL
#               that we used last time, and edit it appropriately:
#
# https://grch37.ensembl.org/Homo_sapiens/Export/Output/Location?db=core;output=fasta;r=20:57410001-57510000;genomic=unmasked;_format=Text
#
#             - save the output as chr20-100kbp.37.fasta
#             - then read it with
mySeq37 <- readFasta("chr20-100kbp.37.fasta")
#
#             - (verifications as before)
#
#             - calculate offset to mySeq position 1
#                 We downloaded from 20:57410001-57510000 - therefore a
#                 position i in myseq is i + 57410000 on chr20.
#
#             - Now we can extract our nucleotides ...
myGenomeRef <- data.frame(START = myIntRef$START,
                          REF = mySeq37[(myIntRef$START - 57410000)],
                          stringsAsFactors = FALSE)
#               ... and compare:
all(myIntRef$REF == myGenomeRef$REF, na.rm = TRUE)
#             Very nice.



# [END]
